OpenNewTab: Support Vivaldi

sctice opened the pull request:
Vivaldi is a Chromium-based browser that has decided to present itself as Chrome but that differs from Chrome in a number of ways. In particular, Vivaldi uses shift+click to open a link in a new foreground tab and ctrl+shift+click to open a link in a new foreground *window*.

This adds support to detect Vivaldi and change the simulated click that the "OpenNewTab" command executes accordingly. Per [this SO answer][1], we must use the extension tabs API to identify Vivaldi, and this requires some extra work that we'd prefer to do just once, so we do it as part of the options setup on the background Program. Additionally, the check requires accessing a property that the WebExtension definition of `browser.tabs.Tab` doesn't know about, so we have to extend that interface in @types.globals.d.ts for this special case.

[1]: https://stackoverflow.com/a/77047611

Closes https://github.com/lydell/LinkHints/issues/39

sctice commented:
I've successfully run `npm test` and tested on Vivaldi. I haven't had time yet to test on Chrome/FF. I think it's unlikely this change will affect FF, but there's a little more risk with Chrome because it would be pretty easy for the Vivaldi check to mistakenly identify Chrome or other Chromium variants as Vivaldi.

lydell commented on src/background/Program.ts:
```diff
@@ -2083,11 +2087,23 @@ export default class BackgroundProgram {
     log.level = options.logLevel;
   }
 
+  async getChromiumVariant(): Promise<string> {
+    if (BROWSER !== "chrome") {
+      return "";
+    }
+    const tabs = await browser.tabs.query({
+      active: true,
+      currentWindow: true,
+    });
+    return tabs?.[0]?.vivExtData !== undefined ? "vivaldi" : "chrome";
+  }
```
Could you move this to a function at the bottom of the file since it never uses `this`?

Would it make sense to run this _every time_ inside the `openNewTab` method, instead of just once at startup basically? I think it should be so fast that it won’t be noticeable. The reason is that it would feel nice to have the Vivaldi stuff localized to the only place where it matters, instead of having `this.chromiumVariant`. If we ever need Vivaldi specific stuff in the future, we can refactor then. You can use the `fireAndForget` helper (which is already used in the method) to avoid `openNewTab` becoming `async`.

sctice replied:
> Could you move this to a function at the bottom of the file since it never uses `this`?

Definitely.




> Would it make sense to run this _every time_ inside the `openNewTab` method, instead of just once at startup basically?

I totally appreciate the desire to keep the necessary changes localized, and I wouldn't mind accommodating that, even though I do think it makes more sense to establish the specific browser once (since it won't be changing) and passing it around where it's needed, but I think the complication with the tab API being asynchronous is a little more sticky unless I'm misunderstanding something.

> You can use the `fireAndForget` helper (which is already used in the method) to avoid `openNewTab` becoming `async`.

I don't see how that will work since we actually need the value wrapped by the promise returned by `browser.tabs.query()`. It looks to me like `fireAndForget()` would only help if we didn't care about resolving the promise. I don't see a way to use the result of `browser.tabs.query()` without making the logic downstream of that also asynchronous/promise-aware. That struck me as a bigger change with more potential implication, which is why I opted to do it in the background program initialization, which seemed to already have other asynchronous work.

If you'd strongly prefer to do the browser detection right before the click, I can try it out.

lydell commented on src/background/Program.ts:
```diff
@@ -2083,11 +2087,23 @@ export default class BackgroundProgram {
     log.level = options.logLevel;
   }
 
+  async getChromiumVariant(): Promise<string> {
```
```suggestion
  async getChromiumVariant(): Promise<"chrome" | "vivaldi"> {
```

This change makes it so that TypeScript yells at us if we try to use any other string than those. Could you use this in the message sent to the worker as well?

lydell reviewed (commented):
Nice, I like it! I left some comments.

sctice commented:
I've pushed some changes to address comments. I went with the approach of detecting Vivaldi "online", right before dispatching the `OpenNewTab` message. It is a more compact change set. I haven't had time to try testing it out with the various ways that `openNewTab()` ends up getting called.

sctice commented:
Fixed a bug I introduced while focusing on opening tabs in the foreground. We _do_ want to ctrl-click in Vivaldi when opening a tab in the background (we just don't want `shift` in that case). In short, ctrl-click and shift-click are mutually exclusive in Vivaldi except in the case that you want to open a new _window_, which we never want to do in the context of handling the `OpenNewTab` message.

sctice commented:
Thanks for working on this with me! I see now what you meant about using `fireAndForget()`. And I like dropping the empty string now that we're only checking in a chrome context.

lydell commented:
Status: Hoping to get some time to merge and make a release soon.
